luigi/contrib/snowflake.py

```diff
@@ -0,0 +1,523 @@
+import abc
+import logging
+
+import luigi
+
+from enum import Enum
+from luigi import six
+from luigi.contrib import postgres
+from luigi.contrib import rdbms
+
+logger = logging.getLogger('luigi-interface')
+
+# Snowflake's library does not have constant for their error code.
+# We define our own
+SNOWFLAKE_TABLE_EXISTS_ERROR = 2002
+SNOWFLAKE_GENERIC_PROGRAMMING_ERROR = 2003
+
+try:
+    import snowflake
+    import snowflake.connector
+except ImportError:
+    logger.warning("Loading snowflake module without snowflake-connector-python installed. "
+                   "Will crash at runtime if snowflake functionality is used.")
+
+
+class SnowflakeSupportedAuth(Enum):
+    default = 0
+    aws_key_auth = 0
+    # next value = 1
+
+    @classmethod
+    def get(cls, which):
+        if which == cls.aws_key_auth:
+            return AwsKeyAuth()
+        else:
+            raise ValueError("SnowflakeSupportedAuth value ' {0} ' isn't supported", which)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class SnowflakeAuth(object):
+    """Abstractable SnowflakeAuth base class that can be replace by tool
+    specific implementation."""
+    @property
+    def configuration_section(self):
+        """
+        Override to change the configuration section used
+        to obtain default credentials.
+        """
+        return 'snowflake'
+
+    @abc.abstractmethod
+    def credentials(self, task):
+        pass
+
+
+class _SettingsMixins(object):
+    """
+    Generic mixins that defines how to fetch the database and warehouse
+    configuration for a given snowflake integration.
+    """
+    @property
+    def configuration_section(self):
+        """
+        Override to change the configuration section used
+        to obtain default credentials.
+        """
+        return 'snowflake'
+
+    @property
+    def warehouse(self):
+        """
+        Fetch the warehouse configuration from the luigi configuration file under:
+
+        [snowflake]
+        warehouse=value
+        """
+        return luigi.configuration.get_config().get(self.configuration_section, 'warehouse', default=None)
+
+    @property
+    def role(self):
+        """
+        Fetch the role configuration from the luigi configuration file under:
+
+        [snowflake]
+        role=value
+        """
+        return luigi.configuration.get_config().get(self.configuration_section, 'role', default=None)
+
+
+class AwsKeyAuth(SnowflakeAuth):
+    def credentials(self):
+        """
+        Return an AWS credential string for the provided task.
+
+        If no valid credentials are set, raise a NotImplementedError.
+        """
+        if self.aws_access_key_id and self.aws_secret_access_key:
```

Comment:
i understand this is still WIP. Does AWS snowflake only accept key-based auth? Does it support role-based?